src/player.cpp

```diff
@@ -4374,10 +4390,64 @@ void player::update_stamina( int turns )
         }
     }
 
-    stamina = roll_remainder( stamina + stamina_recovery * turns );
+    // Decrease stamina regeneration if tired
+    float fatigue_mod = std::max( ( 1 - get_fatigue() / 500.0f ), 0.1f );
+    stamina_recovery *= fatigue_mod;
+
+    // Calculate stamina change and cap at max
+    const int max_stam_penalty = get_stamina_max() - get_stamina_max_penalty();
+    int stamina_change = std::min( roll_remainder( stamina_recovery * turns ),
+                                   max_stam_penalty - stamina );
+    stamina += stamina_change;
+
+    // Add winded effect
+    if( stamina < ( get_stamina_max() / 10 ) ) {
+        add_effect( effect_winded, 1_turns, num_bp, true );
+    } else {
+        remove_effect( effect_winded );
+    }
+
+    // Increase needs with stamina recovery
+    // Sleeping character will recover stamina pool, thus regenerate stamina
+    // This regeneration should not increase needs
+    if( !has_effect( effect_sleep ) ) {
+        int StaminaIncreaseHunger = get_option< int >( "PLAYER_RECOVER_STAMINA_INCREASE_HUNGER" );
+        float mutation_hunger = 1.0f + mutation_value( "metabolism_modifier" );
+        float AddHunger = roll_remainder( stamina_change / 10000.0f * StaminaIncreaseHunger *
+                                          mutation_hunger );
+        if( is_npc() ) {
+            AddHunger *= 0.25f;
+        }
+        mod_stat( "hunger", AddHunger );
+
+        int StaminaIncreaseThirst = get_option< int >( "PLAYER_RECOVER_STAMINA_INCREASE_THIRST" );
+        float mutation_thirst = 1.0f + mutation_value( "thirst_modifier" );
+        float AddThirst = roll_remainder( stamina_change / 10000.0f * StaminaIncreaseThirst *
+                                          mutation_thirst );
+        if( is_npc() ) {
+            AddThirst *= 0.25f;
+        }
+        mod_stat( "thirst", AddThirst );
+
+        int StaminaIncreaseFatigue = get_option< int >( "PLAYER_RECOVER_STAMINA_INCREASE_FATIGUE" );
+        float mutation_fatigue = 1.0f + mutation_value( "fatigue_modifier" );
+        float AddFatigue = roll_remainder( stamina_change / 10000.0f * StaminaIncreaseFatigue *
+                                           mutation_fatigue );
```

Comment:
I'll suggest a lambda function here since its usefulness would probably be limited to just this function.

Something along the lines of:
```c++
auto modify_stat = [&](const std::string& stat, const std::string& option, const std::string& mutation_mod, bool npc_reduce){
    const int option_val = get_option<int>( option );
    float mutation_val = 1.0f + mutation_value( mutation_mod );
    float mod_result = roll_remainder( stamina_change / 10000.0f * option_val * mutation_val );
    if( npc_reduce ) {
        mod_result *= 0.25f;
    }
    mod_stat( stat, mod_result );
};
// Example Call
modify_stat( "fatigue", "PLAYER_RECOVER_STAMINA_INCREASE_FATIGUE", "fatigue_modifier", is_npc() );
```